src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -820,31 +821,91 @@ HRESULT CorProfiler::ProcessReplacementCalls(
       const auto original_argument = pInstr->m_Arg32;
       const void* module_version_id_ptr = &module_metadata->module_version_id;
 
-      // insert the opcode and signature token as
-      // additional arguments for the wrapper method
-      //
-      // IMPORTANT: Conditional branches may jump to this call instruction, so
-      // we cannot add the argument-loading instructions BEFORE this instruction,
-      // otherwise this will surface in an InvalidProgramException.
-      // Either begin loading the additional arguments starting with this ILInstr
-      // structure or make the current ILInstr a no-op and do all of the argument
-      // loading after this instruction.
+      // Begin IL Modification
       ILRewriterWrapper rewriter_wrapper(&rewriter);
       rewriter_wrapper.SetILPosition(pInstr);
+
+      // IL Modification #1: Replace original method call with a NOP, so that all original
+      //                     jump targets resolve correctly and we correctly populate the
+      //                     stack with additional arguments
+      //
+      // IMPORTANT: Conditional branches may jump to the original call instruction which
+      // resulted in the InvalidProgramException seen in
+      // https://github.com/DataDog/dd-trace-dotnet/pull/542. To avoid this, we'll do
+      // the rest of our IL modifications AFTER this instruction.
       auto original_methodcall_opcode = pInstr->m_opcode;
       pInstr->m_opcode = CEE_NOP;
+      pInstr = pInstr->m_pNext;
+      rewriter_wrapper.SetILPosition(pInstr);
 
-      // replace with a non-virtual call (CALL) to the instrumentation wrapper
-      // always use CALL because the wrappers methods are all static
-      rewriter_wrapper.CallMemberAfter(wrapper_method_ref, false);
-      rewriter_wrapper.SetILPosition(pInstr->m_pNext);
+      // IL Modification #2: Conditionally box System.Threading.CancellationToken
+      //                     if it is the last argument in the target method.
+      //
+      // If the last argument in the method signature is of the type
+      // System.Threading.CancellationToken (a struct) then box it before calling our
+      // integration method. This resolves https://github.com/DataDog/dd-trace-dotnet/issues/662
+      // which reproduces when the CLR automatically boxes System.Threading.CancellationToken
```

Comment:
nit: CLR was never boxing the `CancellationToken`.

We were basically the value of  `CancellationToken.m_source` (which is a pointer to a `CancellationTokenSource`) into an `object`, and then casting that into a `CancellationTokenSource`, which should be fine.

I still don't understand why that fails for `System.Data` in x86 but works everywhere else.

Reply:
Thanks for the correction, I was definitely incorrect on the "CLR automatically boxing" the struct, I will fix that.

I don't understand why it is happening, because I even tested this with the HttpMessageHandler integration and that seems to work just fine. I realized what ultimately happened though: Without boxing the `CancellationToken` argument, when we enter the integration method, the values placed in the `opCode` variable and the `CancellationTokenSource` variable are switched! The `opCode` can be a value like `VIRTCALL` which is equal to `0x6F` and so we try to read that value and blow up. I have no explanation for it, but boxing the argument does not result in this weird error.